pkg/pgmodel/pgx.go

```diff
@@ -404,76 +435,230 @@ func (p *pgxInserter) getMetricTableName(metric string) (string, error) {
+	}
+	inserter := h.Sum64() % uint64(len(p.inserters))
+	return p.inserters[inserter]
+}
+
+type InsertHandler struct {
+	conn    pgxConn
+	input   chan insertDataRequest
+	pending orderedMap
+}
+
+type orderedMap struct {
+	elements map[string]*list.Element
+	order    list.List // list of PendingBuffer
+
+	oldBuffers []*PendingBuffer
+}
+
+type PendingBuffer struct {
+	metricTable   string
+	needsResponse []insertDataTask
+	batch         SampleInfoIterator
+	start         time.Time
+}
+
+const (
+	flushSize    = 2000
+	flushTimeout = 500 * time.Millisecond
+)
+
+func runInserterRoutine(conn pgxConn, input chan insertDataRequest) {
+	handler := InsertHandler{
+		conn:    conn,
+		input:   input,
+		pending: makeOrderedMap(),
+	}
+
+	for {
+		if !handler.hasPendingReqs() {
+			stillAlive := handler.waitForReq()
+			if !stillAlive {
+				return
+			}
+			continue
 		}
 
+		startHandling := time.Now()
+	hotReceive:
+		for {
+			for i := 0; i < 1000; i++ {
+				receivingNewReqs := handler.handleMoreReqs()
+				if !receivingNewReqs {
+					break hotReceive
+				}
+			}
+			if time.Now().Sub(startHandling) > flushTimeout {
+				handler.flushTimedOutReqs()
+				startHandling = time.Now()
+			}
+		}
+
+		handler.flushEarliestReq()
 	}
-	inserter = inserterI.(chan insertDataRequest)
+}
 
-	return inserter
+func (h *InsertHandler) hasPendingReqs() bool {
+	return !h.pending.IsEmpty()
 }
 
-func (p *pgxInserter) metricTableInsert(metric string, input chan insertDataRequest) {
-	tableName, err := p.getMetricTableName(metric)
-	if err != nil {
-		return
+func (h *InsertHandler) waitForReq() bool {
+	req, ok := <-h.input
+	if !ok {
+		return false
 	}
 
-	batch := SampleInfoIterator{sampleInfos: make([]*samplesInfo, 0), sampleIndex: -1, sampleInfoIndex: 0}
+	h.handleReq(req)
+
+	return true
+}
+
+func (h *InsertHandler) handleMoreReqs() bool {
+	select {
+	case req := <-h.input:
+		h.handleReq(req)
+		return true
+	default:
+		return false
+	}
+}
 
-	const batchSize int = 1000
+func (h *InsertHandler) handleReq(req insertDataRequest) bool {
+	needsFlush, pending := h.pending.addReq(req)
+	if needsFlush {
+		h.flushPending(pending)
+		return true
+	}
+	return false
+}
 
-	var needsResponse []insertDataTask
+func (p *PendingBuffer) addReq(req insertDataRequest) bool {
+	p.needsResponse = append(p.needsResponse, insertDataTask{finished: req.finished, errChan: req.errChan})
+	p.batch.sampleInfos = append(p.batch.sampleInfos, req.data...)
+	return len(p.batch.sampleInfos) > flushSize
+}
 
+func (h *InsertHandler) flushTimedOutReqs() {
 	for {
-		req, ok := <-input
-		if !ok {
+		earliest := h.pending.Front()
+		if earliest == nil {
 			return
 		}
-		needsResponse = append(needsResponse, insertDataTask{finished: req.finished, errChan: req.errChan})
-		batch.sampleInfos = append(batch.sampleInfos, req.data...)
 
-	TryReadMore:
-		for len(batch.sampleInfos) < batchSize {
+		earliestPending := earliest.Value.(*PendingBuffer)
```

Comment:
nit can we avoid these casts in non-orderedMap code. can Front() return both a *PendingBuffer and a List.Element (which I would call a ref). So the code would be `earliest, earliestRef := h.pending.Front()`